### PR TITLE
bcrypt_ext: Add compatibility with libxcrypt

### PR DESCRIPTION
**Pull request by besser82**
The return values of `crypt()` and `crypt_gensalt()` should be treated as real c-strings instead of binary data by the module.

### CODE REVIEW

**Comment by besser82** [comment]
The failing tests are not releated to this change.

**Comment by jdoss** [comment]
Any ETA on getting this merged? Bcrypt on with `ruby-2.5.0-90.fc28.x86_64` on Fedora 28 Beta when using https://github.com/codahale/bcrypt-ruby/blob/master/lib/bcrypt/password.rb#L46 puts a bunch of null bytes on the end of the output. 

We pointed our Gemfile at https://github.com/besser82/bcrypt-ruby/tree/libxcrypt and it totally fixes the problem.

File: ext/mri/bcrypt_ext.c [hunk 1]
```diff
@@ -18,7 +18,7 @@ static VALUE bc_salt(VALUE self, VALUE prefix, VALUE count, VALUE input) {
 
     if(!salt) return Qnil;
 
-    str_salt = rb_str_new2(salt);
+    str_salt = rb_str_new(salt, strlen(salt));
```
Comment (tenderlove) [inline ext/mri/bcrypt_ext.c]: Is this actually necessary to fix the bug?  It seems like `rb_str_new2` [already does that](https://github.com/ruby/ruby/blob/b768dc2207d800e1c462fc4367e89b2ae2f5a919/string.c#L793-L798). (`rb_str_new2` [is just an alias](https://github.com/ruby/ruby/blob/b768dc2207d800e1c462fc4367e89b2ae2f5a919/include/ruby/intern.h#L835) for `rb_str_new_cstr`)

**Comment by ojab** [comment]
JFYI: Fedora-28 with libxcrypt is released.

**Comment by duritong** [comment]
Fedora 28 was release 2 days ago and at the moment you can't use bcrypt on it. Can we get that fixed? Thanks!

**Comment by ojab** [comment]
Unfortunately it's not a solution for rvm/rbenv/etc-based workflow.

**Comment by tjschuck** [comment]
I have just pushed version 3.1.12.rc1 to RubyGems.org which should solve this problem.

Can someone please confirm that `gem install bcrypt --prerelease` now works on Fedora 28?  If so, I'll push out the final version.

**Comment by ojab** [comment]
Works fine here (F28 w/ all updates, ruby-2.5.0p0 via rvm), thanks.

**Comment by dessel** [comment]
Yes, now everything is ok on F28

**Comment by miry** [comment]
Thanks. It fixes issue with

```
Sorcery::CryptoProviders::BCrypt.encrypt "Welcome2019", "ffLLizQt4eF_Dpd5mwcc"
```

`raw_hash` contains  a huge binary string (sometime to other variables). It looks like `size` was calculated wrong for Fedora 30.

**Pull request merged**